keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+    expression: "object.spec.replicas <= params.maxReplicas"
+    name: max-replicas
+  defaultValidations:
+  - expression: "object.spec.replicas <= 5"
+    enforcement: [Deny]
+```
+
+Reporting/debugging/analysis implications:
+
+- Violations for non-singleton policies will always be reported using a {policy
+  definition, policy binding} identifier pair. To be consistent with this,
+  singleton policies can be use a {policy definition, policy definition}
+  identifier pair. This is a bit verbose but keeps reporting consistent and
+  makes tracing a violation back to resources that produced it straight-forward.
+
+### Phase 2
+
+All these capabilities are required before Beta, but will not be implemented
+in the first alpha release of this enhancement due to the size and complexity
+of this enhancement.
+
+#### Namespace scoped policy binding
+
+For phase 1, policy bindings were only allowed to be cluster scoped. We can
+support namespace scoped policy bindings as follows:
+
+- Add a `NamespacePolicyBinding` resource.
+- If the parameter resource is namespace scoped, it implicitly matches
+  resources only in the namespace it is in, but may further constrain what
+  resources it matches with additional match criteria.
+
+Benefits: Allows policy of a namespace to be controlled from within the
+namespace. As an example, ResourceQuota works this way.
+
+Details to consider:
+
+- Should a policy support both cluster scoped and namespace scoped binding? If
+  so how? It would need two different parameter CRDs (since a CRD must either be
+  cluster scoped or namespace scoped, not both).
+
+#### CEL Expression Composition
+
+##### Variables
+
+<<[UNRESOLVED jpbetz, ?? ]>>
+cel-policy-template "terms" are very convenient. But we don't have consensus on
+if they are sufficiently essential. Need to either agree on something or drop
+this.
+<<[/UNRESOLVED]>>
+
+Each CEL "program" is a single expression. There is no support for vaiable
+assignment. This can result in redundant code to traverse maps/arrays or
+dereference particular fields.
+
+We can support this in much the same way as cel-policy-template if we offer an
+ordered list or variables. These can be lazily evaluated while the validation
+expressions are evaluated (cel-policy-template does this). The results can also
+be memoized to avoid repeated evaluations if they are shared across validations.
+
+```yaml
+  variables:
+    - name: metadataList
+      expression: "spec.list.map(x, x.metadata)"
+    - name: itemMetadataNames
+      expression: "metadataList.map(m, m.name)"
+  validations:
+    - expression: "itemMetadataNames.all(name, name.startsWith('xyz-'))"
+    - expression: "itemMetadataNames.exists(name, name == 'required')"
+```
+
+##### Scopes
+
+<<[UNRESOLVED jpbetz, ?? ]>> 
+This is not yet agreed on, and some reviewers have found it more
+difficult to understand validations using this feature than validations that do
+not.
+
+Also, the "bind varables" idea is maybe too ambitious?
```

Comment:
What makes "bind variables" ambitious? CEL Policy templates already has this behavior.

Reply:
nvm, I see this is binding variables WRT scopes, not variables as defined in the "Variables" section immediately above.

Reply:
I would omit scopes. There has been an ask for them in the past, but usually this signals the need for a separate policy.

Reply:
I've dropped scopes out. I've kept in content in the "alternatives considered" section for reference.